sstables/index_reader.hh

```diff
@@ -186,6 +186,11 @@ public:
     }
 
     processing_result process_state(temporary_buffer<char>& data) {
+        if (_abort != nullptr && _abort->abort_requested()) {
+            sstlog.warn("Aborting index consumer. Reason : %s", _abort->abort_requested_exception_ptr());
+            return proceed::no;
+        }
+
```

Comment:
We can't just stop like that. We have to return something, no? How will the caller proceed? It thinks it needs to wait for I/O, but no I/O is forthcoming.

Reply:
I think we may have to throw here. It seems proceed::no can cause input_stream<CharType>::consume() to end prematurely (silently), which apparently may result in sstable writer generating a bad filter in shutdown (because it (wrongly) thinks the index was entirely consumed). With the exception, we're aborting the process which initiated the consumption of index (e.g. regeneration of summary or filter).

Reply:
It's not the end of the world if we lose a sstable write in shutdown. That happens today with compactions that are almost completed but do get aborted.

Reply:
@raphaelsc That's what I ended up doing. Instead of returning `proceed:no`, the exception_ptr in abort source is now rethrown. I was actually concerned about allowing the exception to propagate up causing the write to fail, so I was trying to handle it in `maybe_rebuild_filter_from_index()`. But what you say makes sense, so I have removed the exception handling and the write fails when the index reader gets aborted.

One thing to note is, `maybe_rebuild_filter_from_index()` is also in the path of memtable flush to disk but although the filter will never be rebuilt during that flow as the filter will always be optimal. So hopefully this is okay.

Reply:
from exception handling perspective, we don't retry flush on shutdown, so we'll have to replay later through commitlog. So we're good here.

As for not stalling flush with the rebuild from index (because that can result in memory pressure. that's what you're concerned right?), maybe it can be a problem with twcs, which segregates memtable data into multiple write streams, one for each window involved. let's say that memtable data spans 2 windows, where 99% of it belongs to a single window. Estimation average the input partition count by number of windows spanned, so we have suboptimal filters for both. Maybe we should bypass maybe_rebuild_filter_from_index() if sstable origin == "memtable", meaning we're favoring flush time (decreasing mem pressure as we're only able to clear memtable after sealing sstable), and sstables produced by flush have a size proportional to memory size, so overestimated filters for them won't play a large role in memory usage, and they can later be adjusted through compactions. what do you think?

Reply:
> from exception handling perspective, we don't retry flush on shutdown, so we'll have to replay later through commitlog. So we're good here.

Nice! This is what I was actually concerned about. Although the chances of a filter rebuild happening during a memtable flush are almost zero, I was concerned about missing any edge cases that I'm not aware of which could produce sub-optimal filters triggering a filter rebuild. It's good to know that commitlog replay will take care of failures during such cases.

Reply:
> As for not stalling flush with the rebuild from index (because that can result in memory pressure. that's what you're concerned right?), maybe it can be a problem with twcs, which segregates memtable data into multiple write streams, one for each window involved. let's say that memtable data spans 2 windows, where 99% of it belongs to a single window. Estimation average the input partition count by number of windows spanned, so we have suboptimal filters for both. Maybe we should bypass maybe_rebuild_filter_from_index() if sstable origin == "memtable", meaning we're favoring flush time (decreasing mem pressure as we're only able to clear memtable after sealing sstable), and sstables produced by flush have a size proportional to memory size, so overestimated filters for them won't play a large role in memory usage, and they can later be adjusted through compactions. what do you think?

This should be totally fine! One question though : with other strategies like LCS and STCS, during memtable flush, the exact partition count is known and is passed to bloom filter when it is created. So they always turn out to be optimal when checked inside `maybe_rebuild_filter_from_index()` and the filter is never rebuilt. IIUC, with TWCS, individual memtables within each windows will be flushed to their own separate sstables. In that case, why do we have to use a partition estimate to create bloom filters? Can't we use the same logic as LCS and STCS to create filters with the exact partition count?

Reply:
> > As for not stalling flush with the rebuild from index (because that can result in memory pressure. that's what you're concerned right?), maybe it can be a problem with twcs, which segregates memtable data into multiple write streams, one for each window involved. let's say that memtable data spans 2 windows, where 99% of it belongs to a single window. Estimation average the input partition count by number of windows spanned, so we have suboptimal filters for both. Maybe we should bypass maybe_rebuild_filter_from_index() if sstable origin == "memtable", meaning we're favoring flush time (decreasing mem pressure as we're only able to clear memtable after sealing sstable), and sstables produced by flush have a size proportional to memory size, so overestimated filters for them won't play a large role in memory usage, and they can later be adjusted through compactions. what do you think?
> 
> This should be totally fine! One question though : with other strategies like LCS and STCS, during memtable flush, the exact partition count is known and is passed to bloom filter when it is created. So they always turn out to be optimal when checked inside `maybe_rebuild_filter_from_index()` and the filter is never rebuilt. IIUC, with TWCS, individual memtables within each windows will be flushed to their own separate sstables. In that case, why do we have to use a partition estimate to create bloom filters? Can't we use the same logic as LCS and STCS to create filters with the exact partition count?

the problem exists because we have a single memtable, which will be flushed into multiple sstables. It's 1:N, not 1:1. for LCS and ICS, we don't have such a segregator on flush, just during compaction.

Reply:
> > > As for not stalling flush with the rebuild from index (because that can result in memory pressure. that's what you're concerned right?), maybe it can be a problem with twcs, which segregates memtable data into multiple write streams, one for each window involved. let's say that memtable data spans 2 windows, where 99% of it belongs to a single window. Estimation average the input partition count by number of windows spanned, so we have suboptimal filters for both. Maybe we should bypass maybe_rebuild_filter_from_index() if sstable origin == "memtable", meaning we're favoring flush time (decreasing mem pressure as we're only able to clear memtable after sealing sstable), and sstables produced by flush have a size proportional to memory size, so overestimated filters for them won't play a large role in memory usage, and they can later be adjusted through compactions. what do you think?
> > 
> > This should be totally fine! One question though : with other strategies like LCS and STCS, during memtable flush, the exact partition count is known and is passed to bloom filter when it is created. So they always turn out to be optimal when checked inside `maybe_rebuild_filter_from_index()` and the filter is never rebuilt. IIUC, with TWCS, individual memtables within each windows will be flushed to their own separate sstables. In that case, why do we have to use a partition estimate to create bloom filters? Can't we use the same logic as LCS and STCS to create filters with the exact partition count?
> 
> the problem exists because we have a single memtable, which will be flushed into multiple sstables. It's 1:N, not 1:1. for LCS and ICS, we don't have such a segregator on flush, just during compaction.

Maybe it's possible to feed an exact estimation with twcs by having a partition count tracker which impl is compaction strategy specific (that lives in memtable) and we feed it into the splitter since we want each stream to know exactly the amount of partitions it will have to consume for creation of the filter. But it may be some significant amount of work and I don't see we doing it any time soon.

Reply:
Okay.

Reply:
Allowing the write fail, propagates the exception and the server fails due to unhandled exception. So, I have updated the PR to drop the old filter only after the new one is built during rebuild. If the rebuild is aborted, the old filter still remains and will be written to disk.